core/src/test/java/com/linecorp/armeria/server/InitiateConnectionShutdownTest.java

```diff
@@ -0,0 +1,242 @@
+        clientBootstrap.channel(TransportType.socketChannelType(CommonPools.workerGroup()));
+        clientBootstrap.handler(new ChannelInitializer<Channel>() {
+            @Override
+            protected void initChannel(Channel ch) throws Exception {
+                final ChannelPipeline p = ch.pipeline();
+                p.addLast(new LoggingHandler());
+                p.addLast(new Http2ConnectionHandlerBuilder()
+                                  .server(false)
+                                  .frameListener(clientListener)
+                                  .validateHeaders(false)
+                                  .gracefulShutdownTimeoutMillis(0)
+                                  .build());
+                p.addLast(new ChannelInboundHandlerAdapter() {
+                    @Override
+                    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+                        if (evt instanceof Http2ConnectionPrefaceAndSettingsFrameWrittenEvent) {
+                            clientSetupFinished.set(true);
+                            ctx.pipeline().remove(this);
+                        }
+                    }
+                });
+            }
+        });
+
+        final ChannelFuture channelFuture = clientBootstrap.connect(
+                goAwayServer.socketAddress(SessionProtocol.H2C));
+        assertTrue(channelFuture.awaitUninterruptibly().isSuccess());
+        clientChannel = channelFuture.channel();
+        await().timeout(Duration.ofSeconds(5)).untilTrue(clientSetupFinished);
+        http2Client = clientChannel.pipeline().get(Http2ConnectionHandler.class);
+    }
+
+    @ParameterizedTest
+    @CsvSource({ "/goaway_async", "/goaway_blocking" })
+    void initiateConnectionShutdownHttp2Async(String path) throws Exception {
+        final AtomicBoolean finished = new AtomicBoolean();
+        clientChannel.eventLoop().execute(() -> {
+            final ChannelHandlerContext ctx = clientChannel.pipeline().firstContext();
+            final Http2Headers headers = getHeaders(path);
+            http2Client.encoder().writeHeaders(ctx, STREAM_ID, headers, PADDING, false, ctx.newPromise());
+            http2Client.flush(ctx);
+        });
+        when(clientListener.onDataRead(any(), anyInt(), any(), anyInt(), anyBoolean())).thenAnswer(
+                invocation -> {
+                    finished.set(true);
+                    return 0;
+                });
+        await().timeout(Duration.ofSeconds(2)).untilTrue(finished);
+        final InOrder inOrder = inOrder(clientListener);
+        inOrder.verify(clientListener, never()).onGoAwayRead(any(ChannelHandlerContext.class),
+                                                             eq(Integer.MAX_VALUE),
+                                                             eq(Http2Error.NO_ERROR.code()),
+                                                             any(ByteBuf.class));
```

Comment:
Could we verify the content of the debug data for all verifications in this class?

Reply:
Hm, not sure how to do that - seems like client releases the buffer so Mockito cannot compare to it:

```


refCnt: 0
io.netty.util.IllegalReferenceCountException: refCnt: 0
	at io.netty.buffer.AbstractByteBuf.ensureAccessible(AbstractByteBuf.java:1454)
	at io.netty.buffer.AbstractByteBuf.checkIndex(AbstractByteBuf.java:1383)
	at io.netty.buffer.AbstractByteBuf.getLong(AbstractByteBuf.java:459)
```

Do you have a suggestion on how to overcome this?

Reply:
How about intercepting `onGoAwayRead()` with `when(...).thenAnswer(...)` like you did for `onDataRead()`? The buffer should still be available at that point. We could call `retain()` on it and then compare later, or just convert the buffer into a string and collect it into a list and then later do assertion against the collected strings.